### PR TITLE
Feature cargo logs query

### LINKED ISSUE #223: Feature: Options to logs cargo
**Describe the solution you'd like**
Use a new query parameter for cargo logs to enable these features:
- since: number as unix tmestamp
- until: number as unix timestamp
- tail: "all" or number (see bollard types)
- timestamps: boolean Add timestamps to every log line
- follow: boolean return a stream or not (optional)

These options should be passed to bollard docker logs function.

We should also add CLI options to use them for example:

```sh
nanocl cargo logs my-cargo -f -t 100 --timestamps
```

### PR DESCRIPTION
**Pull request by CreepyPvP**
closes #223

### CODE REVIEW

**Comment by CreepyPvP** [comment]
Code cov is failing. Should I try to improve it? Also changed default behaviour of cargo logs, follow defaults to false now to match docker log command

File: bin/nanocl/src/commands/cargo.rs [hunk 1]
```diff
@@ -174,8 +174,15 @@ async fn exec_cargo_logs(
   args: &CargoArgs,
   options: &CargoLogsOpts,
 ) -> Result<(), CliError> {
+  let query = CargoLogQuery { 
+    namespace: args.namespace.clone(), 
+    tail: options.tail.to_owned(),
```
Comment (0xle0ne) [inline bin/nanocl/src/commands/cargo.rs]: can you clone instead ?

File: bin/nanocld/src/utils/cargo.rs [hunk 2]
```diff
@@ -887,15 +887,19 @@ pub async fn patch(
...
   let stream = docker_api.logs(
     &format!("{name}.c"),
     Some(LogsOptions::<String> {
-      follow: true,
+      follow: query.follow.unwrap_or(false),
+      timestamps: query.timestamps.unwrap_or(false),
+      since: query.since.unwrap_or(0),
+      until: query.until.unwrap_or(0),
+      tail: query.tail.to_owned().unwrap_or("all".to_string()),
       stdout: true,
```
Comment (0xle0ne) [inline bin/nanocld/src/utils/cargo.rs]: I think we should add stdout and stderr to `CargoLogQuery` aswell

File: bin/nanocld/src/utils/cargo.rs [hunk 3]
```diff
@@ -887,15 +887,19 @@ pub async fn patch(
...
 pub fn get_logs(
   name: &str,
+  query: &CargoLogQuery,
   docker_api: &bollard_next::Docker,
 ) -> Result<impl StreamExt<Item = Result<Bytes, HttpError>>, HttpError> {
   let stream = docker_api.logs(
     &format!("{name}.c"),
     Some(LogsOptions::<String> {
-      follow: true,
+      follow: query.follow.unwrap_or(false),
```
Comment (0xle0ne) [inline bin/nanocld/src/utils/cargo.rs]: You can use `unwrap_or_default()` there instead.
By default boolean and number will be set to `false` and 0

File: bin/nanocld/src/utils/cargo.rs [hunk 4]
```diff
@@ -887,15 +887,19 @@ pub async fn patch(
 
 pub fn get_logs(
   name: &str,
+  query: &CargoLogQuery,
   docker_api: &bollard_next::Docker,
 ) -> Result<impl StreamExt<Item = Result<Bytes, HttpError>>, HttpError> {
   let stream = docker_api.logs(
     &format!("{name}.c"),
     Some(LogsOptions::<String> {
```
Comment (0xle0ne) [inline bin/nanocld/src/utils/cargo.rs]: An helper `From<LogsOptions::String> for CargoLogQuery` should be use there

**Review by 0xle0ne — CHANGES_REQUESTED** [review]

**Comment by 0xle0ne** [comment]
It's normal to loose a bit of CodeCov dont worry about that!
You can try to improove it if you want. But we should create another issue for it i think

**Comment by 0xle0ne** [comment]
Do you want to update the dashboard/app aswell to apply these features once merged ?
It's kinda needed because we cannot output 20k file logs without make the ui laggy

**Comment by CreepyPvP** [comment]
> Do you want to update the dashboard/app aswell to apply these features once merged ? It's kinda needed because we cannot output 20k file logs without make the ui laggy

sure

File: bin/nanocld/src/utils/cargo.rs [hunk 5]
```diff
@@ -887,20 +887,12 @@ pub async fn patch(
 
 pub fn get_logs(
   name: &str,
-  query: &CargoLogQuery,
+  query: CargoLogQuery,
```
Comment (0xle0ne) [inline bin/nanocld/src/utils/cargo.rs]: You should keep the reference there `&CargoLogQuery` and do a `query.clone().into()`

**Review by 0xle0ne — CHANGES_REQUESTED** [review]

File: bin/nanocld/src/utils/cargo.rs (hunk 5 above)
Comment (CreepyPvP) [reply bin/nanocld/src/utils/cargo.rs]: whats the reason for doing so? Would there not always be 2 instances of the object. While when its up for the parent function to decide if a clone should happen there could be only 1 instance

**Review by 0xle0ne — APPROVED** [review]

**Pull request merged**